pkg/networking/backend_sg_provider.go

```diff
@@ -0,0 +1,245 @@
+package networking
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	awssdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	ec2sdk "github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	networking "k8s.io/api/networking/v1beta1"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/aws/services"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	resourceTypeSecurityGroup = "security-group"
+	tagKeyK8sCluster          = "elbv2.k8s.aws/cluster"
+	tagKeySGType              = "elbv2.k8s.aws/type"
+	tagValueBackend           = "backend"
+
+	explicitGroupFinalizerPrefix = "group.ingress.k8s.aws/"
+	implicitGroupFinalizer       = "ingress.k8s.aws/resources"
+
+	sgDescription = "[k8s] Shared Backend SecurityGroup for LoadBalancer"
+)
+
+// BackendSGProvider is responsible for providing backend security groups
+type BackendSGProvider interface {
+	// Get returns the backend security groups to use
+	Get(ctx context.Context) ([]string, error)
+	// Release cleans up the auto-generated backend SG if necessary
+	Release(ctx context.Context) error
+}
+
+// NewBackendSGProvider constructs a new  defaultBackendSGProvider
+func NewBackendSGProvider(clusterName string, backendSGEnabled bool, backendSGs []string, vpcID string,
```

Comment:
i prefer not to pass backendSGEnabled in, 
this component should solely generates a backendSG if required. 
Whether to call this component(decided by backendSGEnabled) should be handled in caller-side.

Reply:
updated the ingress model builder to rely on the enabledBackendSG configuration.

Since there will be multiple modules relying on this configuration, I feel it is cleaner to have the backend SG provider abstract out the configuration as well. In case feature is disabled, the return value would be empty, and callers can take further action based on it. For now, the backend SG provider doesn't take in the configuration - it is up to the caller to decide.